test/test_client_async.py

```diff
@@ -9,67 +11,55 @@
 
 logging.basicConfig(level=logging.WARNING)
 
+pytestmark = pytest.mark.asyncio
+
+
 ip = '127.0.0.1'
 tcpport = 1102
 db_number = 1
 rack = 1
 slot = 1
 
 
-class TestClient(unittest.TestCase):
-
-    @classmethod
-    def setUpClass(cls):
-        cls.process = Process(target=mainloop)
-        cls.process.start()
-        time.sleep(2)  # wait for server to start
-
-    @classmethod
-    def tearDownClass(cls):
-        kill(cls.server_pid, 1)
-
-    def setUp(self):
-        self.client = snap7.client.Client()
-        self.client.connect(ip, rack, slot, tcpport)
-
-    def tearDown(self):
-        self.client.disconnect()
-        self.client.destroy()
-
-    @unittest.skip("TODO: RuntimeWarning: coroutine 'TestClient.test_as_db_read' was never awaited")
-    async def test_as_db_read(self):
-        size = 40
-        start = 0
-        db = 1
-        data = bytearray(40)
-        self.client.db_write(db_number=db, start=start, data=data)
-        result = await self.client.as_db_read(db_number=db, start=start, size=size)
-        self.assertEqual(data, result)
-
-    @unittest.skip("TODO: RuntimeWarning: coroutine 'TestClient.test_as_db_write' was never awaited")
-    async def test_as_db_write(self):
-        size = 40
-        data = bytearray(size)
-        check = await self.client.as_db_write(db_number=1, start=0, data=data)
-        self.assertEqual(check, 0)
+@pytest.fixture(scope='module')
+def testserver():
+    process = Process(target=mainloop)
+    process.start()
+    time.sleep(2)  # wait for server to start
+    yield process
+    kill(process.pid, 1)
 
-    @unittest.skip("TODO: crash client: FATAL: exception not rethrown")
-    async def test_as_ab_read(self):
-        start = 1
-        size = 1
-        await self.client.as_ab_read(start=start, size=size)
 
-    @unittest.skip("TODO: not yet fully implemented")
-    async def test_as_ab_write(self):
-        start = 1
-        size = 10
-        data = bytearray(size)
-        await self.client.as_ab_write(start=start, data=data)
+@pytest.fixture(scope='module')
+def testclient():
+    client_async = snap7.client_async.ClientAsync()
+    client_async.connect(ip, rack, slot, tcpport)
+    yield client_async
+    client_async.disconnect()
+    client_async.destroy()
 
-    async def test_as_db_get(self):
-        await self.client.db_get(db_number=db_number)
 
-    @unittest.skip("TODO: RuntimeWarning: coroutine 'TestClient.test_as_download' was never awaited")
-    async def test_as_download(self):
-        data = bytearray(128)
-        await self.client.as_download(block_num=-1, data=data)
+async def test_wait_loop(testclient):
```

Comment:
so what I dont like about this weird pytest fixture is that this `testclient` arguments comes out of nowhere, you have no way to quickly see where it comes from. I guess pytest keeps an internal storage of named fixtures you can use? How can you quickly jump to the function definition?

Reply:
I understand what you mean. It isn't marked as pytest_testclient or similiar to see directly what and where it might come from (well you can name them as these) and depending on the IDE (if even used), this may be confusing. I agree with that.
 
There are two ways (I know) how to organise pytest fixtures (stuff you want to reuse in multiple places):
1) They are already existing in the test_file_XYZ itself, marked with a `@pytest.fixture(scope="funciton/module/class/session")`, depending on your testing purposes.

2) conftest.py - pytest is looking for this file in the directory structure. It is a convention in pytest. Each conftest.py provides configs for the file tree pytest finds it in. You can store the wanted fixtures (and other pytest related stuff) in it. So in our case one conftest.py file in the root directory would be enough. 

3) You can mark individual tests with `@pytest.mark.usefixtures('testclient')` (I learned today). So the confusion would be reduced about that. 
Or you can mark all tests in a file with `pytestmark = pytest.mark.webtest`.
If you need multiple marks for all tests: `pytestmark = [pytest.mark.webtest, pytest.mark.asyncio]`

So there are places existing to look for, if you know that pytest is used. It is handy to avoid boilerplate, especially when you want to reuse/parametrise tests.